db.go

```diff
@@ -54,6 +69,91 @@ func (e *Entry) GetAttr(attr string) (Attr, bool) {
 	return v, ok
 }
 
+// Authenticate checks that the LDAP Entry is valid for authentication and that
+// the entry has a password that matches the given password. If the entry is
+// valid for authentication and the password matches the one in the entry, nil
+// is returned. Otherwise an error is returned.
+//
+// The form of hashed password in an entry is described by [RFC 2307, 5.3]
+// Interpreting user and group entries with the supported schemes listed in the
+// [hashSchemes] map above. This is how [OpenLDAP passwords] are done, so this
+// does it the same. e.g. "{SSHA}<base64-encoded-hash+salt>"
+//
+// [RFC 2307, 5.3]: https://datatracker.ietf.org/doc/html/rfc2307#section-5.3
+// [OpenLDAP passwords]: https://www.openldap.org/faq/data/cache/347.html
+func (e *Entry) Authenticate(password string) error {
+	if !e.isValidForAuthentication() {
+		return ErrInvalidEntryForAuth
+	}
+
+	// hashSchemes is a list of supported password hashing schemes
+	// with their functions for checking a password hash against
+	// a password.
+	hashSchemes := map[string]func() hash.Hash{
+		"{SSHA}":    sha1.New,
+		"{SSHA256}": sha256.New,
+		"{SSHA512}": sha512.New,
+	}
+
+	var firstErr error
+	hashedPasswords, _ := e.GetAttr("userPassword")
+	for _, hashedPassword := range hashedPasswords.Vals {
+		for scheme, newHash := range hashSchemes {
```

Comment:
sorry, naive question - it seems like you should be able to 
`index := strings.IndexRune(hashedPassword, "}"`, take substring and lookup?
Or is this too brittle for you?

Reply:
The format is `{scheme}<base64-encoded-hash>`. Just checking for a `}` does not ensure it conforms to that format. I've tried a few things here (I even tried `fmt.Sscanf` but that is different to the C version and seems a bit useless), and this is where I started and ended.

Reply:
Thinking about this some more - you are right. That should be sufficient as it will fail the map lookup if malformed. I'll do it. Thanks

Reply:
done.